shared/common-adapters/search-filter.js

```diff
@@ -0,0 +1,317 @@
+}
+
+type Props = {|
+  icon?: ?IconType,
+  focusOnMount?: ?boolean,
+  negative?: ?boolean,
+  onChange: (text: string) => void,
+  placeholderText: string,
+  style?: ?Styles.StylesCrossPlatform,
+  type: 'Small' | 'Full-width' | 'Mobile',
+  waiting?: ?boolean,
+
+  onBlur?: ?() => void,
+  // If onClick is provided, this component won't focus on click. User is
+  // expected to handle actual filter/search in a separate component, perhaps
+  // in a popup.
+  onClick?: ?() => void,
+  // following props are ignored when onClick is provided
+  hotkey?: ?('f' | 'k'), // desktop only
+  onKeyDown?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+  onKeyUp?: (event: SyntheticKeyboardEvent<>, isComposingIME: boolean) => void,
+|}
+
+type State = {|
+  focused: boolean,
+  hover: boolean,
+  text: string,
+|}
+
+class SearchFilter extends React.PureComponent<Props, State> {
+  state = {
+    focused: false,
+    hover: false,
+    text: '',
+  }
+
+  _input = React.createRef()
+  _onBlur = () => {
+    this.setState({focused: false})
+    this.props.onBlur && this.props.onBlur()
+  }
+  _onFocus = () => {
+    this.setState({focused: true})
+  }
+
+  _focus = () => {
+    if (this.state.focused) {
+      // Can't just make a null for Kb.ClickableBox as that'd cause
+      // PlainInput to be re-constructed.
+      return
+    }
+    this._input.current && this._input.current.focus()
+  }
+  _blur = () => {
+    this._input.current && this._input.current.blur()
+  }
+  _clear = () => {
+    this._update('')
+  }
+  _cancel = e => {
+    this._blur()
+    this._clear()
+    e && e.stopPropagation()
+  }
+  _update = text => {
+    this.setState({text})
+    this.props.onChange(text)
+  }
+  _mouseOver = () => this.setState({hover: true})
+  _mouseLeave = () => this.setState({hover: false})
+  _onHotkey = cmd => {
+    this.props.hotkey && cmd.endsWith('+' + this.props.hotkey) && this._focus()
+  }
+  _onKeyDown = (e: SyntheticKeyboardEvent<>, isComposingIME: boolean) => {
+    e.key === 'Escape' && !isComposingIME && this._cancel()
+    this.props.onKeyDown && this.props.onKeyDown(e, isComposingIME)
+  }
+  _typing = () => this.state.focused || !!this.state.text
+
+  componentDidMount() {
+    this.props.focusOnMount && this._focus()
+  }
+  render() {
+    if (
+      (this.props.type === 'Mobile' && !Styles.isMobile) ||
+      (this.props.type !== 'Mobile' && Styles.isMobile)
+    ) {
+      return (
+        <Kb.Text type="Header" style={{color: Styles.globalColors.red}}>
+          Unsupported type
+        </Kb.Text>
+      )
+    }
+    const iconColor = this.props.negative ? Styles.globalColors.white_75 : Styles.globalColors.black_50
+    const iconSizeType = this.props.type === 'Full-width' ? 'Default' : 'Small'
+    const hotkeyText =
+      this.props.hotkey && !this.props.onClick && !Styles.isMobile
+        ? ` (${Platforms.shortcutSymbol}+${this.props.hotkey.toUpperCase()})`
+        : ''
+    const content = (
```

Comment:
It might be nice to split this up into helper functions since it's pretty big. Not react components, just functions.